src/Castle.Windsor/MicroKernel/DefaultKernel.cs

```diff
@@ -725,7 +726,14 @@ protected virtual IHandler WrapParentHandler(IHandler parentHandler)
 				return null;
 			}
 
-			var handler = new ParentHandlerWrapper(parentHandler, Parent.Resolver, Parent.ReleasePolicy);
+			// If we are disposing child containers with parent registered transients, 
+			IReleasePolicy releasePolicy;
+			if (parentHandler.ComponentModel.LifestyleType == LifestyleType.Singleton || parentHandler.ComponentModel.LifestyleType == LifestyleType.Undefined)
+				releasePolicy = Parent.ReleasePolicy;
+			else
+				releasePolicy = this.ReleasePolicy;
```

Comment:
This will break many scenarios, if the component was registered and resolved by the parent container it might not be a transient, what about pooled, scoped or any custom lifestyle. The component should follow and be tracked by the release policy of the container it was registered in, however understand this isn't really the desired behaviour for transients.

Unfortunately the fix for this won't be so easy, as you can see child containers haven't really been designed.

Windsor however has followed the rule that if you call Resolve you must also call Release even for transients, which would fix the unit test.